torch/_dynamo/output_graph.py

```diff
@@ -1015,3 +1060,51 @@ def create_graph_input(self, name, type_expr=None, before=False):
             else:
                 self.input_name_to_proxy[name] = proxy
             return proxy
+
+    def is_name_bound(self, name):
+        if name in self.input_name_to_proxy:
+            return True
+        for proxy in self.seen_proxies:
+            if proxy.node.name == name:
+                return True
+        return False
+
+    def lift_freevar_to_input(self, proxy):
+        self.create_graph_input(proxy.node.name)
+        self.lifted_freevars.add(proxy)
+        if self.parent is not None and not self.parent.is_name_bound(proxy.node.name):
+            self.parent.lift_freevar_to_input(proxy)
+
+
+# NOTE: [HigherOrderOperator tracing design]
+# Ignoring HigherOrderOperators for a moment,
+# OutputGraph represents the graph being built by Dynamo that may be compiled
+# and executed. It holds a root SubgraphTracer where the FX graph is built.
+#
+# HigherOrderOperators are operators that take functions (represented by
+# torch.fx.GraphModule) as their arguments. When Dynamo encounters a
```

Comment:
I'm trying to understand what `(represented by torch.fx.GraphModule)` implies. Does it mean the functions must be `torch.fx.GraphModule` to begin with? But that doesn't feel right given it is mentioned later that "dynamo introspect the function and capture into GraphModule".

Reply:
Basically, a higher order operator is something like `higher_order_op(f, args)`, where `f` is an arbitrary user-provided function. The problem we're dealing with is - what if `f` depends on things like global variables or any other value that isn't directly passed in to `args`?

Well, we want to lift them to arguments. And luckily, this is exactly what Dynamo can do :) It can find all values that we depend upon (regardless of whether they're local variables, closed over variables, or globals), and lift them to be explicitly inputs to the fx.GraphModule.

So, for example, if you write
```
y = ...
def f(x):
    return x + y
 
higher_order_op(f, inp)
```
we will automatically rewrite that into (more or less)
```
def f_graph(x, y):
    return x + y
higher_order_op(f_graph, inp, y)
```

Reply:
We've got a bit of a vocabulary problem that needs to be resolved at some point. It will probably like the following:
- there is a set of higher order operators that accept arbitrary callables as body functions
- there is a set of higher order operators that accept only sanitized FX graphs (in the form of GraphModules) as inputs. 

For example, you may have functorch.experimental.map(f, xs) that can accept arbitrary functions as inputs, but after Dynamo captures `f` into `f_graph_safe`, it should also rewrite the call to `map` into a call to `torch._ops.map_sanitized(f_graph_safe, xs)`.

We might even have a third category of higher order operators depending on how we decide to handle nn.Modules: nn.Modules could be in the body function or we could ban them from the body function.